File: src/ua/edu/hneu/oop/java/lr2p2/task2/WordFinder.java
package ua.edu.hneu.oop.java.lr2p2.task2;

import java.util.HashSet;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class WordFinder {

    private static final Pattern WORD_PATTERN = Pattern.compile("(?muU)(((^|\\W)[eyuioa]\\w*)|(\\w*[eyuioa]($|\\W)))");

    public Set<String> findAllVowelStartEndWords(String input) {
        Matcher matcher = WORD_PATTERN.matcher(input);
        Set<String> words = new HashSet<>();
        while (matcher.find()) {
            words.add(matcher.group(1).trim());
        }
        return words;
    }
}

File: src/ua/edu/hneu/oop/dotNet/LK2Part2/Task2/WordFinder.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace Task2
{
    public static class WordFinder
    {
        private static readonly Regex WordPattern = new Regex(@"\b[eyuioa]\w*[eyuioa]\b", RegexOptions.Compiled | RegexOptions.CultureInvariant);

        public static HashSet<string> FindAllVowelStartEndWords(this string input)
        {
            HashSet<string> words = new HashSet<string>();
            foreach (Match match in WordPattern.Matches(input))
                words.Add(match.Value);

            return words;
        }
    }
}